### ISSUE TITLE
FoundationApi wrapper for components

### ISSUE THREAD

**Issue by AntJanus** [zurb/foundation-apps#390]
I was thinking about it and FoundationApi is pretty damn low-level for users to use directly in their controllers. It can be sometimes confusing having to explain the pub/sub system and most users don't need that functionality nor most of what comes *with* FoundationApi. It's more of a power-user service than everyday user service.

Here's an example of how to activate a module:

````
foundationApi.publish('module-id', 'show');
````

It looks simple but for new users, they might wonder, "why publish?", and have to remember that's it's "show" or "hide" or any other keyword. And as far as coding goes, it's somewhat of a misdirection.

So I had this idea for some syntactic sugar, it's basically an adapter that hides implementation details and exposes utility methods. Here's what I had in mind:

- an easier-to-understand service that requires `FoundationApi`. Not sure what to call it, maybe just `foundation`? But that'd be confusing
- it would include methods such as `activate`, `toggle`, and `deactivate` which would use foundationApi to send the `foundationApi.publish` method.
- it could have a `getMediaQueries` or even `matchMediaQuery` methods instead of having to manually extracting those from `getSettings`. This could also enable us to expose other functionality.
- it could also include further sugar for adding new components to multi-component directives such as tabs and notifications. 
- it could also include sugar for element factories such as `adapter.create('modal', { config goes here })`. Again, down the road.

There are definitely a ton of advantages to this process:

1. FoundationApi could change in nature whenever it wants to with the adapter as long as the adapter keeps the original sugar methods in place. This means that heavier refactors are totally fine to use.
2. it's easier for new users to understand `adapter.activate('my-modal-id')`. In fact, we could even add sugar to support full elements like `adapter.activate(myModalElement)` somewhere down the road.
3. It's much easier to understand for new users and it'd be the cleanest implementation.
4. it prevents users from using FoundationApi too much. It has utilities that really should be used only by powerusers and for internal use such as `generateUuid`, `animate`, and so on.

The other awesome thing is that we can introduce this adapter right away and get new users using it while supporting old users as well. Any eventually deprecation of public `FoundationApi` methods could happen down the road in BC updates which wouldn't break anything for the adapter users.

As a great comparison, it reminds me of the Symfony vs Laravel MVC approach where Symfony forces you to do manual DI, exposes lots of internal methods, requires you to do all kinds of complicated stuff while Laravel employs Facades and similar approaches to hide the complexity and give users more flexibility and readable code.

**Comment by circlingthesun** [comment]
Having used other angular UI modules, its indeed not very intuitive needing a ```foundationApi``` service is for component specific actions. What makes most sense to me is exposing component specific functionality in a separate service. For instance a ```fzModal``` service could simply have an ```activate``` member. Then as you suggest  ```foundationModal``` could depend on ```foundationApi```. The same would apply for say ```zfNotifications```.

**Comment by AntJanus** [comment]
Hmm, I'm not sure if I'm understanding correctly. Are you talking about building out a separate service per module that could activate/deactivate them instead of a service that can activate any module?

I'm also working on a factory for modals that would allow us to create modals programmatically (rather than with separate markup) and that one *will* have an activate/deactivate member built in; however, it will have the ability to be activated/deactivated via FoundationApi as well. Is that more along the lines of what you're thinking?

**Comment by circlingthesun** [comment]
Yeah, that's exactly what I meant. Sorry for being unclear. Exposing components in a separate service seems like a much more angular way of doing things. The way things are now feels reminiscent to the way things work in Foundation 5/JQuery. Just my 2c.

**Comment by AntJanus** [comment]
Hmm.. I'm not sure if I agree with creating a separate service for each module. I kind of like the idea of having a central adapter that activates/deactivates components and then the components themselves handle how that happens. 

This is so that the existing implementation of components and their activation stays intact and still available to new users. It also follows the way the components talk to each other behind the scenes.

However, I think that once I'm done with the prototype ModalFactory, you'll see something more angular-like behavior with the factory creating the component directives and controlling its activation state.

**Comment by circlingthesun** [comment]
Well, you could still dependency inject the central service into each component service and keep the way components talk to each other. For backwards compatibility I can see why you would prefer keeping the current architecture. However if you intend to create a proxy for communicating with the central service, perhaps its a good idea to create a decentralised facade and give every component its own proxy.

**Comment by AntJanus** [comment]
Hmm. Alright. That's a fair point. I'll prototype something out and see how it looks.

**Comment by tunecino** [comment]
When worked with bootstrap's modal before i remember having used a provider for it :
<pre>
.provider('modalState', function($stateProvider) {
    var provider = this;
    this.$get = function() {
        return provider;
    }
    this.state = function(stateName, options) {
        var modalInstance;
        $stateProvider.state(stateName, {
            url: options.url,
            onEnter: function($modal, $state) {
                modalInstance = $modal.open(options);
                modalInstance.result['finally'](function() {
                    modalInstance = null;
                    if ($state.$current.name === stateName) {
                        $state.go('^');
                    }
                });
            },
            onExit: function() {
                if (modalInstance) {
                    modalInstance.close();
                }
            }
        });
    };
})
</pre>

Then on my router configs I just had to to use it as a page wrapper when needed like this : 
<pre>
modalStateProvider
    .state('app.campaign.new', {
         url: '/new',
         templateUrl: 'partials/campaign/campaign.new.html',
         controller: 'CampaignCreateController'
    });
</pre> 
It was elegant as solution to me while It meant to satisfy a Restful CRUD app where each page had it's own controller and i needed that easy access with a <code>/new</code> link any way. 
Now with Foundation for apps doing almost the same CRUD thing, i just assigned my Edit page to a parent one and wrapped it into modal then replaced Cancel and Close buttons with go back links : http://foundation.zurb.com/forum/posts/23322-open-apps-modal-on-page-load?page=2#comment-21214
So I'm sharing this just to point to one more possible use of modals, I mean why not add an optional wrapper parameter to pages like : 
<pre>
---
name: edit
url: /edit
parent: items
controller: EditlCtrl
wrapper: modal
---
</pre>
where page will be initialized on a modal with its own controller or maybe even sharing it's parent controller and $scope.
